File: 91-decode-ways/decode-ways.py
class Solution:
    def solve(self, s: str, index: int, memo: dict[int, int]) -> int:
        if index in memo:
            return memo[index]
        if index == len(s):
            return 1
        if s[index] == '0':
            return 0

        ways = self.solve(s, index+1, memo)
        if (index+1 < len(s)) and (s[index] == '1' or (s[index] == '2' and s[index+1] <= '6')):
            ways += self.solve(s, index+2, memo)

        memo[index] = ways
        return memo[index]


    def numDecodings(self, s: str) -> int:
        memo = {}
        return self.solve(s, 0, memo)

File: 91-decode-ways/decode-ways.cpp
class Solution {
public:
    int solve(string s, int index, vector<int> &memo){
        if (index > s.length()) return 0;
        if (index == s.length()) return 1;
        if (s[index] == '0') return 0;
        if (memo[index] != -1) return memo[index];

        int ways = solve(s, index + 1, memo);
        if (index < s.length() && (s[index] == '1' || (s[index] == '2' && s[index + 1] <= '6'))){
            ways += solve(s, index + 2, memo);
        }

        memo[index] = ways;
        return ways;
    }
    int numDecodings(string s) {
        vector<int> memo(s.length(), -1);
        return solve(s, 0, memo);
    }
};